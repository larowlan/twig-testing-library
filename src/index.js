import { getQueriesForElement, prettyDOM } from "@testing-library/dom"
import Twig from "twig"
import fs from "fs"
import "regenerator-runtime/runtime"
import DrupalAttribute from "drupal-attribute"

const mountedContainers = new Set()

if (typeof afterEach === "function") {
  afterEach(cleanup)
}

async function render(
  twigFile,
  context = {},
  namespaces = {},
  twigCallback = () => {}
) {
  const baseElement = document.body
  const container = baseElement.appendChild(document.createElement("div"))

  // Add it to the mounted containers to cleanup.
  mountedContainers.add(container)
  twigCallback(Twig)

  container.innerHTML = await loadTemplate(twigFile, context, namespaces)

  return {
    container,
    baseElement,
    debug: (el = baseElement, maxLength, options) =>
      Array.isArray(el)
        ? // eslint-disable-next-line no-console
          el.forEach((e) => console.log(prettyDOM(e, maxLength, options)))
        : // eslint-disable-next-line no-console,
          console.log(prettyDOM(el, maxLength, options)),
    ...getQueriesForElement(baseElement),
  }
}

const loadTemplate = async (file, context = {}, namespaces) => {
  Twig.registryReset = () => {
    Twig.Templates.registry = {}
  }

  Twig.cache(false)
  Twig.extendFunction("create_attribute", (value) => new DrupalAttribute(value))
  Twig.twigAsync = (options) => {
    return new Promise((resolve, reject) => {
      options.load = resolve
      options.error = reject
      options.async = true
      options.autoescape = false
      options.namespaces = namespaces

      if (options.data || options.ref) {
        try {
          resolve(Twig.twig(options))
        } catch (error) {
          reject(error)
        }
      } else {
        fs.readFile(options.path, "utf8", (err, data) => {
          if (err) {
            reject(new Error(`Unable to find template file ${options.path}`))
            return
          }
          options.load = (template) => {
            template.rawMarkup = data
            resolve(template)
          }
          Twig.twig(options)
        })
      }
    })
  }
  return Twig.twigAsync({
    path: file,
  }).then((template) => {
    if (!context.hasOwnProperty("attributes")) {
      context.attributes = new DrupalAttribute()
    }
    return template.render(context)
  })
}

function cleanup() {
  mountedContainers.forEach(cleanupContainer)
}

function cleanupContainer(container) {
  if (container.parentNode === document.body) {
    document.body.removeChild(container)
  }
  mountedContainers.delete(container)
}

// just re-export everything from dom-testing-library
export * from "@testing-library/dom"
export { render, cleanup, Twig }

/* eslint func-name-matching:0 */
