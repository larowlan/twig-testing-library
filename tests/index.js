/* eslint no-new: 0 */
import Accordion from './fixtures/accordion';
import {render, fireEvent, Twig} from "../src";

Twig.extendFilter("backwords", (text) => {
  return text.split(" ").reverse().join(" ");
});

describe('Test library by testing an accordion', () => {
  it('Can be initially rendered open', async () => {
    const { container, getByText } = await render('./tests/fixtures/accordion.twig', {
      // This is intentionally backwards so we can test extending twig.
      title: 'title Accordion',
      open: true,
    }, {
      'twig-testing-library-tests': './tests/fixtures/'
    });
    const accordionElement = container.querySelector('.accordion');
    const summaryElement = accordionElement.querySelector('summary');
    const accordion = new Accordion.Accordion(accordionElement);
    accordion.init();
    expect(accordionElement).toMatchSnapshot('Initial render');
    expect(accordionElement.classList.contains('accordion--open')).toBe(true);
    fireEvent.click(getByText('Accordion title'));
    expect(accordionElement).toMatchSnapshot('First collapse');
    expect(accordionElement.classList.contains('accordion--open')).toBe(false);
    expect(summaryElement.getAttribute('aria-expanded')).toEqual('false');
    expect(summaryElement.getAttribute('aria-pressed')).toEqual('false');
    fireEvent.click(getByText('Accordion title'));
    expect(accordionElement).toMatchSnapshot('Re-open');
    expect(accordionElement.classList.contains('accordion--open')).toBe(true);
    expect(summaryElement.getAttribute('aria-expanded')).toEqual('true');
    expect(summaryElement.getAttribute('aria-pressed')).toEqual('true');
  });

  it('Can be initially rendered closed', async () => {
    const { container, getByText } = await render('./tests/fixtures/accordion.twig', {
      // This is intentionally backwards so we can test extending twig.
      title: 'title Accordion',
      open: false,
    }, {
      'twig-testing-library-tests': './tests/fixtures/'
    });
    const accordionElement = container.querySelector('.accordion');
    const summaryElement = accordionElement.querySelector('summary');
    const accordion = new Accordion.Accordion(accordionElement);
    accordion.init();
    expect(accordionElement).toMatchSnapshot('Initial render');
    expect(accordionElement.classList.contains('accordion--open')).toBe(false);
    fireEvent.click(getByText('Accordion title'));
    expect(accordionElement).toMatchSnapshot('First expand');
    expect(accordionElement.classList.contains('accordion--open')).toBe(true);
    expect(summaryElement.getAttribute('aria-expanded')).toEqual('true');
    expect(summaryElement.getAttribute('aria-pressed')).toEqual('true');
    fireEvent.click(getByText('Accordion title'));
    expect(accordionElement).toMatchSnapshot('Re-collapse');
    expect(accordionElement.classList.contains('accordion--open')).toBe(false);
    expect(summaryElement.getAttribute('aria-expanded')).toEqual('false');
    expect(summaryElement.getAttribute('aria-pressed')).toEqual('false');
  });
});
