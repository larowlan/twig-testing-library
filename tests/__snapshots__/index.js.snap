// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test library by testing an accordion Can be initially rendered closed: First expand 1`] = `
<details
  class="accordion accordion--open"
  open=""
>
  
  
  <summary
    aria-expanded="true"
    aria-pressed="true"
    class="accordion__title"
  >
    Accordion title
  </summary>
  
  
  <div
    class="accordion__content"
  >
    
          
    <p>
      Cras quis nulla commodo, aliquam lectus sed, blandit augue. Cras ullamcorper bibendum bibendum. Duis tincidunt urna non pretium porta. Nam condimentum vitae ligula vel ornare. Phasellus.

    </p>
    
      
  </div>
  

</details>
`;

exports[`Test library by testing an accordion Can be initially rendered closed: Initial render 1`] = `
<details
  class="accordion"
>
  
  
  <summary
    class="accordion__title"
  >
    Accordion title
  </summary>
  
  
  <div
    class="accordion__content"
  >
    
          
    <p>
      Cras quis nulla commodo, aliquam lectus sed, blandit augue. Cras ullamcorper bibendum bibendum. Duis tincidunt urna non pretium porta. Nam condimentum vitae ligula vel ornare. Phasellus.

    </p>
    
      
  </div>
  

</details>
`;

exports[`Test library by testing an accordion Can be initially rendered closed: Re-collapse 1`] = `
<details
  class="accordion"
>
  
  
  <summary
    aria-expanded="false"
    aria-pressed="false"
    class="accordion__title"
  >
    Accordion title
  </summary>
  
  
  <div
    class="accordion__content"
  >
    
          
    <p>
      Cras quis nulla commodo, aliquam lectus sed, blandit augue. Cras ullamcorper bibendum bibendum. Duis tincidunt urna non pretium porta. Nam condimentum vitae ligula vel ornare. Phasellus.

    </p>
    
      
  </div>
  

</details>
`;

exports[`Test library by testing an accordion Can be initially rendered open: First collapse 1`] = `
<details
  class="accordion"
>
  
  
  <summary
    aria-expanded="false"
    aria-pressed="false"
    class="accordion__title"
  >
    Accordion title
  </summary>
  
  
  <div
    class="accordion__content"
  >
    
          
    <p>
      Cras quis nulla commodo, aliquam lectus sed, blandit augue. Cras ullamcorper bibendum bibendum. Duis tincidunt urna non pretium porta. Nam condimentum vitae ligula vel ornare. Phasellus.

    </p>
    
      
  </div>
  

</details>
`;

exports[`Test library by testing an accordion Can be initially rendered open: Initial render 1`] = `
<details
  class="accordion accordion--open"
  open=""
>
  
  
  <summary
    class="accordion__title"
  >
    Accordion title
  </summary>
  
  
  <div
    class="accordion__content"
  >
    
          
    <p>
      Cras quis nulla commodo, aliquam lectus sed, blandit augue. Cras ullamcorper bibendum bibendum. Duis tincidunt urna non pretium porta. Nam condimentum vitae ligula vel ornare. Phasellus.

    </p>
    
      
  </div>
  

</details>
`;

exports[`Test library by testing an accordion Can be initially rendered open: Re-open 1`] = `
<details
  class="accordion accordion--open"
  open=""
>
  
  
  <summary
    aria-expanded="true"
    aria-pressed="true"
    class="accordion__title"
  >
    Accordion title
  </summary>
  
  
  <div
    class="accordion__content"
  >
    
          
    <p>
      Cras quis nulla commodo, aliquam lectus sed, blandit augue. Cras ullamcorper bibendum bibendum. Duis tincidunt urna non pretium porta. Nam condimentum vitae ligula vel ornare. Phasellus.

    </p>
    
      
  </div>
  

</details>
`;
